### ISSUE TITLE
Redact hostnames and IP addresses from files included in debug.zip

### ISSUE THREAD

**Issue by fabiog1901** [cockroachdb/cockroach#87487]
**Is your feature request related to a problem? Please describe.**
Customer's policy dictates that hostnames and IP addresses should not be shared externally. When creating a debug.zip file, we should give an option to redact such fields.

**Describe the solution you'd like**
We can enhance our redaction facility by adding redaction marks around IP addresses and hostnames. Then, we can using some simple hashing function to mask the real names (including the hostname saved in the file name). With hashing, we are still keeping the correlation so we can still make sense of the flow between nodes.

Example: from 

```
lgvase12v.banking.com: sending this to there
abcde.banking.com: pulling this over here
152.185.45.52
```

to (using md5sum, truncated to first 8 chars)

```
540e7005: sending this to there
ce986d82: pulling this over here
ip-fb9d7246  <--- ip-<md5sum>
```

[Slack conversation](https://cockroachlabs.slack.com/archives/CHVV403F0/p1661957124720959)

Jira issue: CRDB-19369

gz#14028

Epic CRDB-20791

**Comment by knz** [comment]
I have investigated this further. the issue is not trivial and it's not so clear that just 1 team can make progress on this. It seems to be a multi-team effort.

There are two separate problems to solve:

1) inclusion of hostnames / IP addresses in specific fields where they are clearly identified as such. Here is a non-exhaustive list:

   - the hostname prefix in log *file names*
   - the hostname / address field in node status records (included in e.g. debug zip `nodes.json`)
   - the hostname / address payload inside range descriptors (included in `rangelog.json`)

   (Note that we **currently do not know the full list of such fields** and the knowledge about them is currently distributed across all CockroachDB engineering teams.)
 
2) inclusion of hostnames / IP addresses in fields where they are mixed with other data. **In that case they are not identified as such** and we cannot perform redaction easily while retaining the other data around them. Here is a non-exhaustive list:

   - output log events (both structured and unstructured)
   - error messages produced while receiving debug zip if a node is temporarily unavailable
   - copy of error messages inside job descriptors (e.g. retrieved in debug zip `jobs.json`)
   - copy of error messages in other persisted data that we use for troubleshooting

   A naive reader may believe we have two sub-problems here but really it's just one problem:
   - hostname / IP addresses when included inside redaction markers. In that case it's easy to mistakenly believe we can do something about them, however sadly we use the same redaction markers both for hostnames / addresses, and other text. So we can't build a heuristic that says "redact only the hostnames / addresses but keep the rest".
   - hostname / IP addresses when mixed with other text without redaction markers. In that case we truly have no metadata to tell us what to redact.

   Both sub-problems boil down to the same: **we have not built infrastructure that identify hostnames / IP addresses when mixed together with other data in text output**  and so we cannot redact them separately from the rest using a simple heuristic.

Bottom line: **we cannot apply a heuristic on our end without a deep, cross-team effort**.  A sophisticated solution here would be a major undertaking.

(But see my next message for a suggestion)

**Comment by knz** [comment]
- As my previous message explains, building a general-purpose solution that "intelligently recognizes and redacts out hostnames and IP addresses" is untractable in the short term. It would require deep effort from many engineering teams. There's also more design context [here](https://docs.google.com/document/d/1IOQpicWc0grb-3VdirDDZoZphsFeecZxAsyUZxhrBuA/edit) (internal doc).

- However, we can perhaps do something less sophisticated. If **the customer would provide a list of strings they don't want to see**, say, as a configuration flag (e.g. `cockroach --forbidden-strings=this.host.name,that.host.name,1.2.3.4`) then we could use this and do search-and-replace on our end in the various API outputs. We could even do this at a very low level to ensure that the substitution is performed even when individual engineering teams forget to do it in their code.

  This approach would guarantee that the forbidden strings are not present in the output **but it would come at another cost**: the search-and-replace might replace *too many strings* and over-correct by replacing things that *look like* hostnames / IP addresses but are not. For example if we have internal troubleshooting numeric data like `10.30.40.50.60.70` if we redact `10.30.40.50` the resulting string `redacted.60.70` might be invalid.  This might reduce our ability to automate investigations in certain cases, but the customer might still be OK with it.

**Comment by bdarnell** [comment]
> So we can't build a heuristic that says "redact only the hostnames / addresses but keep the rest".

Is that the goal? It's unclear from the original message; my interpretation of it is that some customers want to redact hostnames/addresses *in addition to* everything else we already redact, while it sounds like you're saying the customer wants to *always* redact hostnames/IPs even if everything else is unredacted. That seems like a deeply silly policy if you ask me, but there's no arguing with some of these policies. (The linked doc is also talking about treating hostnames as less sensitive than other redacted data).

The original post also talks about replacing the redacted data with a hash, which is unlike what we currently do for redaction. Is that also a customer requirement, or is it just a suggestion that the customer will permit if we find it useful? Hashing IP addresses is fairly useless; it's trivial to brute force a hash if you know it's an ipv4 address (especially if you can limit it to the RFC 1918 private IP blocks). 

Anyway, my first choice would be to treat hostnames and IPs as redactable in exactly the same way as everything else. This doesn't require any fundamentally new capabilities, just an audit to make sure that all the places where hostnames show up are correctly getting marked as redactable. Customers with a strict policy against exposing hostnames would be unable to send us unredacted debug zips; I'm OK with that because I don't want to receive this customer's unredacted data anyway and it's our responsibility to make sure that our debug information is useful even when redacted.

If we really want to engage further with this feature request, given the ambiguity about whether hostnames are more or less sensitive than other redacted data, it seems we'd have no choice but to build a system for categorizing different pieces of redactable data and turning individual categories on and off. 

I'm skeptical of the regex filter approach. It's easy to recognize ipv4 addresses in standard format (and I'm not concerned about e.g. an ipv4 address being logged as a single int32), but hostnames are challenging. If the customer has a good grasp of their DNS hierarchy they may be able to write a regex that matches their hostnames, but in general I would expect that to be difficult.

**Comment by fabiog1901** [comment]
Just to clarify on the request:

- customer wants to redact hostnames/addresses _in addition to_ everything else we already redact.
- the idea behind hashing hostnames/IPs was to make sure the same hostname/IP results in the same hash, thus preserving the correlation. An alternative is to keep a hashmap of hostname:redacted-hostname pairs. 
- If we simply redact every hostname/IP with `<redacted>` we lose the correlation, which is important for troubleshooting: 

Example:

```
host1 sends message to host2
host2 responds to host1
```

should translate to 

```
redacted-1 sends message to redacted-2
redacted-2 responds to redacted-1
```

instead of simply

```
<redacted> sends message to <redacted>
<redacted> responds to <redacted>
```

**Comment by knz** [comment]
Fabio, regarding this:
> customer wants to redact hostnames/addresses in addition to everything else we already redact.

This is the part that is nonsensical / problematic. It seems to me that hostnames/addresses are *already* redacted during our redaction process. Do we have evidence that is not the case?

**Comment by knz** [comment]
(cc @abarganier -  this is important)

@fabiog1901 provides the following items that clarify why debug zip is not redacting information properly:

- the content of generated `nodes.json` as well as any other payload retrieved via a RPC instead of SQL, contain PII even when redaction is enabled. **This is because we did not create proper fine-grained redaction rules for the RPC data retrieval like we did for SQL.**
- `rangelog.json` also contains PII because it contains copies of range descriptors without redaction markers. We need to implement fine-grained redaction for rangelog data.
- the log file names contain the hostname. When redaction is enabled, the log files should be renamed.

**Comment by abarganier** [comment]
Thanks for the callout @knz. 

I'm working with the team to come up with a plan.

**Comment by bdarnell** [comment]
> the idea behind hashing hostnames/IPs was to make sure the same hostname/IP results in the same hash, thus preserving the correlation. An alternative is to keep a hashmap of hostname:redacted-hostname pairs.

We have a solution for this already in our telemetry system: instead of hashing things, encrypt (or hmac) them with `cluster.secret`. This is a value that is shared across the cluster but is specifically excluded from the systems that collect data to send to CRL, so that we can't recover the original data and don't have a hash that can potentially be brute-forced. 

> If we simply redact every hostname/IP with <redacted> we lose the correlation, which is important for troubleshooting:

How important is this in practice? The same could be said about a lot of redacted data, but in general we just remove the redacted data instead of encrypting it, since even the encrypted form leaks a little bit of information (length, cardinality, etc). 

It's not unusual for one side of the connection to log a hostname while the other logs an ip address, which breaks the correlation. In these cases you'd need access to the raw data and the ability to resolve hostnames to see if they match. 

We also try to log node ids instead of hostnames or ip addresses as much as possible, so it's often possible to do the necessary debugging with non-redacted node ids instead of potentially sensitive information.

**Comment by fabiog1901** [comment]
> We also try to log node ids instead of hostnames or ip addresses as much as possible, so it's often possible to do the necessary debugging with non-redacted node ids instead of potentially sensitive information.

Indeed this would be the ideal solution: if we were to solely use node_id's as identifies for the cluster nodes, then we can safely just use `<redacted-host>` for all remaining hostnames/addresses (which represents clients, external endpoints, etc..) and forget about the need for keeping a map or encrypt/hash as we have no correlation (important for troubleshooting) to preserve.
